### DISCUSSION TITLE
How to type a config item that can be both a list or dict?

### DISCUSSION THREAD

**Discussion by pawamoy** [mkdocs/mkdocs#3417] [Q&A] [answered]
On Python 3.10 it looks like we can do `Type(list | dict)`. This won't work on Python 3.8 and 3.9.

**Comment by pawamoy** [comment]
I suppose we have to subclass `Type`, like `Plugins` does:

```python
class Plugins(OptionallyRequired[plugins.PluginCollection]):
    def run_validation(self, value: object) -> plugins.PluginCollection:
        if not isinstance(value, (list, tuple, dict)):
            raise ValidationError('Invalid Plugins configuration. Expected a list or dict.')
```

So something like:

```python
class ListOrDict(Type[Union[list, dict]]):
    def run_validation(self, value: object) -> ...:
        if not isinstance(value, (list, tuple, dict)):
            raise ValidationError('Invalid configuration. Expected a list or dict.')
```

**Comment by oprypin** [comment, marked as answer]
`Type((list, dict))` is a supported way to write this.

https://github.com/mkdocs/mkdocs/blob/00b648f0884244a139f034554b59a032113176f6/mkdocs/config/config_options.py#L331

---

`Type(list | dict)` works by accident only on newer versions of Python at run time
<details><summary> but does not pass mypy:</summary>

```
 error: No overload variant of "Type" matches argument type "UnionType"  [call-overload]
 note: Possible overload variants:
 note:     def [T] __init__(self, type_: type[T], length: int | None = ..., **kwargs: Any) -> Type[T]
 note:     def [T] __init__(self, type_: tuple[type[T], ...], length: int | None = ..., **kwargs: Any) -> Type[T]
```

</details>

---

Note also that if you find yourself allowing both `list | dict`, the config could probably use a more specific implementation that's much more useful for type checking.
For example:

```python
class ListOfItemsThatCanBeWrittenAsDict(BaseConfigOption[List[MyItem]]):
    def run_validation(self, value: object) -> plugins.PluginCollection:
        if not isinstance(value, (list, dict)):
            raise ValidationError('Invalid configuration. Expected a list or dict.')
        if isinstance(value, dict):
            return [MyItem(name=k, **v) for k, v in value.items()]
        else:
            return [MyItem(**v) for v in value]
```

As a further note, `MyItem` here could itself probably use a sub-implementation as a `Config`, otherwise this code is still missing a lot of error checking implementation. 
For example, see here how to implement an item that's fully type checked and should be specified as a dict but can alternatively be written out as just one string instead
https://github.com/mkdocs/mkdocs/blob/00b648f0884244a139f034554b59a032113176f6/mkdocs/config/config_options.py#L948

**Reply by pawamoy** [reply]
Ah, I was sure there was a way to do this (I had seen this previously) but couldn't find any examples in the docs or my own projects! I missed the overload and the tuple with ellipsis in MkDocs' code :sweat_smile: 

Thanks for the reminder and for the detailed code examples :slightly_smiling_face:
